From fbb7a503b807a35eecf365680a5e49212241d939 Mon Sep 17 00:00:00 2001
From: Alessandro Decina <alessandro.decina@collabora.co.uk>
Date: Wed, 14 Sep 2011 16:19:08 +0200
Subject: [PATCH 069/154] pvrvideosink: make the sink work with BGRx

Make the sink work with BGRx so we can do some testing. Add yuv later on.
---
 sys/pvr2d/gstpvrvideosink.c |   31 ++++++++++++++++++++++---------
 1 files changed, 22 insertions(+), 9 deletions(-)

diff --git a/sys/pvr2d/gstpvrvideosink.c b/sys/pvr2d/gstpvrvideosink.c
index 99c2285..5abb8d6 100644
--- a/sys/pvr2d/gstpvrvideosink.c
+++ b/sys/pvr2d/gstpvrvideosink.c
@@ -30,6 +30,8 @@
 /* Debugging category */
 #include <gst/gstinfo.h>
 
+#include <gst/video/video.h>
+
 #define DEFAULT_QUEUE_SIZE 12
 #define DEFAULT_MIN_QUEUED_BUFS 1
 
@@ -65,10 +67,8 @@ static GstStaticPadTemplate gst_pvrvideosink_sink_template_factory =
 GST_STATIC_PAD_TEMPLATE ("sink",
     GST_PAD_SINK,
     GST_PAD_ALWAYS,
-    GST_STATIC_CAPS ("video/x-raw-yuv-strided, "
-        "framerate = (fraction) [ 0, MAX ], "
-        "width = (int) [ 1, MAX ], " "height = (int) [ 1, MAX ]")
-    );
+    GST_STATIC_CAPS (GST_VIDEO_CAPS_BGRx)
+);
 
 enum
 {
@@ -102,7 +102,7 @@ gst_pvrvideosink_get_dcontext (GstPVRVideoSink * pvrvideosink)
   gint refresh_rate;
   DRI2WSDisplay *displayImpl;
   WSEGLError glerror;
-  WSEGLCaps *glcaps;
+  const WSEGLCaps *glcaps;
 
   dcontext = g_new0 (GstDrawContext, 1);
   dcontext->p_blt_info = 0;
@@ -210,7 +210,16 @@ gst_pvrvideosink_create_window (GstPVRVideoSink * pvrvideosink, GstCaps * caps)
       pfnWSEGL_GetDrawableParameters (dcontext->drawable_handle, &source_params,
       &render_params);
   dcontext->dst_mem = (PVR2DMEMINFO *) render_params.hPrivateData;
+
   dcontext->p_blt_info->pDstMemInfo = dcontext->dst_mem;
+  dcontext->p_blt_info->CopyCode = PVR2DROPcopy;
+  dcontext->p_blt_info->BlitFlags = PVR2D_BLIT_DISABLE_ALL;
+
+  dcontext->p_blt_info->DstSurfWidth = video_width;
+  dcontext->p_blt_info->DstSurfHeight = video_height;
+  dcontext->p_blt_info->DstStride =
+      gst_video_format_get_row_stride (GST_VIDEO_FORMAT_BGRx, 0, video_width);
+  dcontext->p_blt_info->DstFormat = PVR2D_ARGB8888;
 
   GST_DEBUG_OBJECT (pvrvideosink, "end");
   return xwindow;
@@ -251,7 +260,6 @@ gst_pvrvideosink_blit (GstPVRVideoSink * pvrvideosink, GstBuffer * buffer)
     return;
   }
 
-
   if (video_width > dcontext->display_width)
     video_width = dcontext->display_width;
   if (video_height > dcontext->display_height)
@@ -265,9 +273,14 @@ gst_pvrvideosink_blit (GstPVRVideoSink * pvrvideosink, GstBuffer * buffer)
   dcontext->p_blt_info->pSrcMemInfo = dcontext->src_mem;
   dcontext->p_blt_info->SrcSurfWidth = video_width;
   dcontext->p_blt_info->SrcSurfHeight = video_height;
-  dcontext->p_blt_info->SrcFormat = PVR2D_YUV422_YUYV;
-  dcontext->p_blt_info->SrcFormat |= PVR2D_SURFACE_PDUMP;
-  dcontext->p_blt_info->SrcStride = 4096;
+  dcontext->p_blt_info->ColourKey = CKEY_MASK_8888;
+  dcontext->p_blt_info->SrcSurfHeight = video_height;
+  dcontext->p_blt_info->SrcSurfWidth = video_width;
+  dcontext->p_blt_info->SrcFormat = PVR2D_ARGB8888;
+  dcontext->p_blt_info->SrcStride = video_width * 4;
+  dcontext->p_blt_info->SrcStride =
+      gst_video_format_get_row_stride (GST_VIDEO_FORMAT_BGRx, 0, video_width);
+
   dcontext->p_blt_info->SrcX = 0;
   dcontext->p_blt_info->SrcY = 0;
   dcontext->p_blt_info->SizeX = video_width;
-- 
1.7.5.4

